File: Android/shared/src/main/kotlin/shared/dto/objects/Salary.kt
@file:UseSerializers(
    UUIDSerializer::class,
    DateISOSerializer::class
)
package shared.dto.objects

import kotlinx.serialization.Serializable
import kotlinx.serialization.UseSerializers
import shared.dto.enums.PaymentType
import shared.dto.enums.SalaryPaymentType
import shared.dto.primitives.Price
import shared.dto.primitives.Wage
import shared.dto.protocols.Parametable
import shared.serializers.UUIDSerializer
import shared.serializers.DateISOSerializer
import java.util.Date
import java.util.UUID

object Salary {
    data object Parameters {
        data object Rules {
            @Serializable
            data class Create(
                val percent: Int?,
                val grid: Map<UUID, SalaryPaymentType>?,
                val wage: Wage,
            ) : Parametable()
        }

        data object Balance {
            @Serializable
            data class Payout(
                val paymentType: PaymentType,
                val dateTo: Date,
            ) : Parametable()

            @Serializable
            data class Calculate(
                val dateTo: Date,
            ) : Parametable()
        }
    }

    data object Responses {

        data object Helpers {

            @Serializable
            data class ServiceWithPaymentType(
                val service: Service.Responses.Partial,
                val paymentType: SalaryPaymentType
            )
        }

        data object Rules {
            @Serializable
            data class Full(
                val percent: Int? = null,
                val grid: List<Helpers.ServiceWithPaymentType>? = null,
                val wage: Wage? = null,
            ) : Parametable()
        }

        data object Balance {
            @Serializable
            data class Full(
                val wage: Price? = null,
                val grid: List<Price>? = null,
                val percent: List<Price>? = null,
                val sum: List<Price>? = null,
            ) : Parametable()
        }
    }
}

File: Node/src/dto/objects/Salary.ts
import { SalaryPaymentType } from "../enums/SalaryPaymentType";
import { Int } from "../tsPrimitives/Int";
import { Wage } from "../primitives/Wage";
import { Price } from "../primitives/Price";
import { UUID } from "../tsPrimitives/UUID";
import { Service } from "./Service";
import { ValidateNested } from "class-validator";
import { IsoDate } from "../tsPrimitives/IsoDate";
import { PaymentType } from "../enums/PaymentType";

export namespace Salary {
    
    export namespace Parameters {
        
        export namespace Rules {

            export class Create {
                @ValidateNested()
                percent?: Int
                
                @ValidateNested()
                grid?: Map<UUID, SalaryPaymentType>

                @ValidateNested()
                wage?: Wage

                constructor(
                    percent?: Int,
                    grid?: Map<UUID, SalaryPaymentType>,
                    wage?: Wage,
                ) {
                    this.percent = percent
                    this.grid = grid
                    this.wage = wage
                }
            }
        }

        export namespace Balance {

            export class Payout {
                paymentType: PaymentType
                dateTo: IsoDate

                constructor(
                    paymentType: PaymentType,
                    dateTo: IsoDate
                ) {
                    this.paymentType = paymentType
                    this.dateTo = dateTo
                }
            }

            export class Calculate {
                dateTo: IsoDate

                constructor(
                    dateTo: IsoDate
                ) {
                    this.dateTo = dateTo
                }
            }
        }
    }
    
    export namespace Responses {

        export namespace Helpers {

            export class ServiceWithPaymentType {
                @ValidateNested()
                service: Service.Responses.Partial

                @ValidateNested()
                paymentType: SalaryPaymentType

                constructor(
                    service: Service.Responses.Partial,
                    paymentType: SalaryPaymentType
                ) {
                    this.service = service
                    this.paymentType = paymentType
                }
            }
        }
        
        export namespace Rules {
            
            export class Full {
                @ValidateNested()
                percent?: Int

                @ValidateNested()
                grid?: Helpers.ServiceWithPaymentType[]

                @ValidateNested()
                wage?: Wage

                constructor(
                    percent?: Int,
                    grid?: Helpers.ServiceWithPaymentType[],
                    wage?: Wage,
                ) {
                    this.percent = percent
                    this.grid = grid
                    this.wage = wage
                }
            }
        }

        export namespace Balance {
            
            export class Full {
                @ValidateNested()
                wage?: Price

                @ValidateNested()
                grid?: Price[]

                @ValidateNested()
                procent?: Price[]

                @ValidateNested()
                sum: Price[]

                constructor(
                    sum: Price[],
                    wage?: Price,
                    grid?: Price[],
                    procent?: Price[],
                ) {
                    this.wage = wage
                    this.grid = grid
                    this.procent = procent
                    this.sum = sum
                }
            }
        }
    }
}